### PR TITLE
Dynamic bat

### PR DESCRIPTION
**Pull request by degasus**
This is a fork of PR #1882, with a few small fixup commits.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4146)

### CODE REVIEW

**Comment by degasus** [comment]
@magumagu Are you fine with my additional patches?

**Comment by JMC47** [comment]
On top of all of the fixes mentions in the other pull request, We Love Golf! and Summer Athletics 2009 are now booting.  They were dcbz sensitive games, so, make up your own conclusion on what fixed them.

File: Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp [hunk 1]
```diff
@@ -368,43 +368,37 @@ void Jit64::dcbz(UGeckoInstruction inst)
...
-  AND(32, R(RSCRATCH2), Imm8(~31));
-  PXOR(XMM0, R(XMM0));
-  MOVAPS(MComplex(RMEM, RSCRATCH2, SCALE_1, 0), XMM0);
-  MOVAPS(MComplex(RMEM, RSCRATCH2, SCALE_1, 16), XMM0);
-
-  // Slow path: mask the address, then call the general-case code.
-  SwitchToFarCode();
-  SetJumpTarget(slow);
-  AND(32, R(RSCRATCH2), Imm32(~31));
   MOV(32, M(&PC), Imm32(jit->js.compilerPC));
```
Comment (phire) [inline Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp]: We don't actually need to do this, clear cache line doesn't use the program counter.

**Comment by JMC47** [comment]
Tested a bunch of games using Dynamic BATs as my main Dolphin build today and everything seems to be fine.

**Comment by lioncash** [comment]
Review status: 0 of 19 files reviewed at latest revision, 8 unresolved discussions.

---

*[Source/Core/Core/HW/Memmap.cpp, line 229 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuV38ZVGv6WVln5ufw:-KPuV38_MaQX_WbfBqe2:b-jeoo2i) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/HW/Memmap.cpp#L229)):*
> ```cpp
> }
> 
> void UpdateLogicalMemory(u32* dbat_table)
> ```

This parameter can be a const pointer.

---

*[Source/Core/Core/HW/Memmap.cpp, line 236 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuUlcftJBFMntL3eSF:-KPuUlcg89pGKEU55mCf:blzqtge) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/HW/Memmap.cpp#L236)):*
> ```cpp
>   }
>   logical_mapped_entries.clear();
>   for (unsigned i = 0; i < (1 << (32 - PowerPC::BAT_INDEX_SHIFT)); ++i)
> ```

These can just be `u32` to remain consistent with all the other variables nearby.

---

*[Source/Core/Core/HW/Memmap.cpp, line 244 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVFPuXqU19Qph5kUw:-KPuVFPuXqU19Qph5kUx:b-z10bwi) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/HW/Memmap.cpp#L244)):*
> ```cpp
>       unsigned logical_size = 1 << PowerPC::BAT_INDEX_SHIFT;
>       unsigned translated_address = dbat_table[i] & ~3;
>       for (PhysicalMemoryRegion& physical_region : physical_regions)
> ```

This can be a const reference as far as I can tell.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 103 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVMWJHlDyiBMI4DZn:-KPuVMWJHlDyiBMI4DZo:bfmw2zg) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L103)):*
> ```cpp
>   } result;
>   u32 address;
>   bool Success() { return result <= PAGE_TABLE_TRANSLATED; }
> ```

This can be a const member function.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 158 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVSj_VApbgFKcXwb-:-KPuVSj_VApbgFKcXwb0:b-v5042t) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L158)):*
> ```cpp
> }
> 
> u32 dbat_table[1 << (32 - BAT_INDEX_SHIFT)];
> ```

std::array

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 159 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVVgWBDEl7K8JaB7O:-KPuVVgWBDEl7K8JaB7P:b-5h08c7) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L159)):*
> ```cpp
> 
> u32 dbat_table[1 << (32 - BAT_INDEX_SHIFT)];
> u32 ibat_table[1 << (32 - BAT_INDEX_SHIFT)];
> ```

same here.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 395 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVfo3F0LxPO_tvNp3:-KPuVfo3F0LxPO_tvNp4:b-iv622s) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L395)):*
> ```cpp
> 
>   u32 hex;
>   // TODO: Refactor this.
> ```

This should likely explain why it should be refactored, to provide context for anyone willing to do it.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 1186 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuW-TBgYL8o_5zOXQv:-KPuW-TBgYL8o_5zOXQw:b-7tplx5) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L1186)):*
> ```cpp
> static void UpdateFakeMMUBat(u32* bat_table, u32 start_addr)
> {
>   for (unsigned i = 0; i < (0x10000000 >> BAT_INDEX_SHIFT); ++i)
> ```

`u32`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

File: Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp (hunk 1 above)
Comment (degasus) [inline Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp]: Also thought so, but hey, DSI :P

**Comment by degasus** [comment]
Review status: 0 of 19 files reviewed at latest revision, 8 unresolved discussions.

---

*[Source/Core/Core/HW/Memmap.cpp, line 229 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuV38ZVGv6WVln5ufw:-KPv5YKj-E9dw0OxTDfN:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/HW/Memmap.cpp#L229)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This parameter can be a const pointer.
</details>
Done.

---

*[Source/Core/Core/HW/Memmap.cpp, line 236 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuUlcftJBFMntL3eSF:-KPv5dDSUeRjedJ9-8VA:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/HW/Memmap.cpp#L236)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> These can just be `u32` to remain consistent with all the other variables nearby.
</details>
Done.

---

*[Source/Core/Core/HW/Memmap.cpp, line 244 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVFPuXqU19Qph5kUw:-KPv5jHVQxNbOLJwJ6ay:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/HW/Memmap.cpp#L244)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This can be a const reference as far as I can tell.
</details>
Done.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 103 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVMWJHlDyiBMI4DZn:-KPv5tc7MRh-hHA6rDoe:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L103)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This can be a const member function.
</details>
Done.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 158 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVSj_VApbgFKcXwb-:-KPv9rGS--Zw_XSbEWpN:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L158)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> std::array
</details>
Done.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 159 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVVgWBDEl7K8JaB7O:-KPv9ro1kmjbWd_ahNH9:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L159)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> same here.
</details>
Done.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 395 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuVfo3F0LxPO_tvNp3:-KPv6GTMPHpyWJQM1tex:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L395)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This should likely explain why it should be refactored, to provide context for anyone willing to do it.
</details>
Done.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 1186 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KPuW-TBgYL8o_5zOXQv:-KPv6OHvPgztk0LhQ5Hg:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8166c7db599d9948b0e62bae25a904e652150757/Source/Core/Core/PowerPC/MMU.cpp#L1186)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `u32`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by JMC47** [comment]
Because Twilight Princess doesn't suffer any kind of performance penalty from MMU being enabled (at least on my PC,) I've marked the dropped sign bug crash as fixed here.

**Comment by Ebola16** [comment]
This PR has the same .elf loading issue as PR #1882 

I just downloaded a build of this PR from https://dl.dolphin-emu.org/prs/pr-4146-dolphin-latest-x64.7z
to see if it has any effect on https://bugs.dolphin-emu.org/issues/8864 (SSBB Alloc Buffer failed).

When attempting to load Gecko OS's .elf Dolphin will crash (attached below). This problem does not occur for me on other builds of Dolphin (tested up to 5.0-443).

Win 10 Pro x64
Intel Core i7-4702MQ CPU @2.20GHz
GeForce GT 750M
2x 8GB Hynix 1600MHz
GeForce Game Ready Driver 372.54
capture
[1SSBB Gecko.zip](https://github.com/dolphin-emu/dolphin/files/427776/1SSBB.Gecko.zip)

**Comment by phire** [comment]
Reviewed 6 of 20 files at r1, 3 of 6 files at r3.
Review status: 9 of 19 files reviewed at latest revision, 9 unresolved discussions.

---

*[Source/Core/Core/HW/Memmap.cpp, line 103 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQ-BM16VxzlpJ12I2fa:-KQ-BM16VxzlpJ12I2fb:b-51p5xn) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f898c5034afbbdf8c06adbca8c8ea63c2b1c1090/Source/Core/Core/HW/Memmap.cpp#L103)):*
> ```cpp
> enum
> {
>   MV_FAKE_VMEM = 1,
> ```

MV Prefix (Memory View) is now incorrect as they are now used as flags for Physical Memory Regions. 

Make this a named enum instead? Move it inside the PhysicalMemoryRegion struct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by phire** [comment]
Reviewed 7 of 20 files at r1, 3 of 6 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[Source/Core/Core/HW/Memmap.cpp, line 233 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQ15WdlYVk2JNXPQT34:-KQ15WdlYVk2JNXPQT35:b-ul33gd) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f898c5034afbbdf8c06adbca8c8ea63c2b1c1090/Source/Core/Core/HW/Memmap.cpp#L233)):*
> ```cpp
>   for (auto& entry : logical_mapped_entries)
>   {
>     g_arena.ReleaseView(entry.mapped_pointer, entry.mapped_size);
> ```

Should be be destroying and recreating the all the mappings for every single BAT change? Would it make sense to scan for entries which have changed? 

mmaping isn't the cheapest of operations.

---

*[Source/Core/Core/PowerPC/PowerPC.h, line 288 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQ17QOXQgkRLsHYuqLx:-KQ17QOXQgkRLsHYuqLy:b51j016) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f898c5034afbbdf8c06adbca8c8ea63c2b1c1090/Source/Core/Core/PowerPC/PowerPC.h#L288)):*
> ```objc
> 
> static const int BAT_INDEX_SHIFT = 17;
> using BatTable = std::array<u32, 1 << (32 - BAT_INDEX_SHIFT)>;
> ```

Could I request a "// 128KB" comment, so I don't have to pull out my calculator next time.

Assuming I even did the math correctly

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by degasus** [comment]
Review status: 17 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[Source/Core/Core/HW/Memmap.cpp, line 103 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQ-BM16VxzlpJ12I2fa:-KQMEJvLItuvPoQQZRMI:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f898c5034afbbdf8c06adbca8c8ea63c2b1c1090/Source/Core/Core/HW/Memmap.cpp#L103)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> MV Prefix (Memory View) is now incorrect as they are now used as flags for Physical Memory Regions. 
> 
> Make this a named enum instead? Move it inside the PhysicalMemoryRegion struct?
</details>
Done.

---

*[Source/Core/Core/HW/Memmap.cpp, line 233 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQ15WdlYVk2JNXPQT34:-KQM8zOoWUVI-O897D5c:b4hqd1x) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f898c5034afbbdf8c06adbca8c8ea63c2b1c1090/Source/Core/Core/HW/Memmap.cpp#L233)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Should be be destroying and recreating the all the mappings for every single BAT change? Would it make sense to scan for entries which have changed? 
> 
> mmaping isn't the cheapest of operations.
</details>
It should, but I'd let this open for the next PR. Especially as no game is known to reconfigure the BATs more than once.

---

*[Source/Core/Core/PowerPC/PowerPC.h, line 288 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQ17QOXQgkRLsHYuqLx:-KQM9YDFO7PVUgusFahE:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f898c5034afbbdf8c06adbca8c8ea63c2b1c1090/Source/Core/Core/PowerPC/PowerPC.h#L288)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Could I request a "// 128KB" comment, so I don't have to pull out my calculator next time.
> 
> Assuming I even did the math correctly
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by degasus** [comment]
The last commit should fix ELF booting, but I haven't verified if our default mapping matches well.

---

Review status: 17 of 20 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-:-KQMJ88vhj7Cw1k8w7l-:b-j0cxge)*

**Comment by JMC47** [comment]
Elf booting is not fully fixed.  The DOL version of Gecko OS works, but the HBC version does not, unless I've already booted the dol version then it does for some reason.

**Comment by lioncash** [comment]
Review status: 17 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 756 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQMInUldSXZ444fgEB4:-KQMInUldSXZ444fgEB5:bbc3n07) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L756)):*
> ```cpp
> 
>   // Translate address
>   u32 bat_result = dbat_table[address >> BAT_INDEX_SHIFT];
> ```

Note that this'll look pretty unusual to anyone new to the code (or anyone who has a grasp for on it, I guess), considering we have a `TranslateAddress` function, which, if inadequate as is, could just have an overload that accepts a BatTable and an address.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 778 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQMJJKy2XyikmGjXaP3:-KQMJJKzYMF_GYfB0fr-:buo9nmt) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L778)):*
> ```cpp
> 
>   // Translate address
>   u32 bat_result = dbat_table[address >> BAT_INDEX_SHIFT];
> ```

Same here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by JMC47** [comment]
@Ebola16 Can you try Gecko OS?  I'm having inconsistencies getting it to work.

**Comment by JMC47** [comment]
SORRY!  I compiled it myself and it works fine now.  Whatever.

LGTM

**Comment by Ebola16** [comment]
.elf files now load again!

**Comment by phire** [comment]
Reviewed 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/HW/Memmap.cpp, line 233 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQ15WdlYVk2JNXPQT34:-KQNS7jmEKMHf4cg7IbO:b-y6etik) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f898c5034afbbdf8c06adbca8c8ea63c2b1c1090/Source/Core/Core/HW/Memmap.cpp#L233)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> It should, but I'd let this open for the next PR. Especially as no game is known to reconfigure the BATs more than once.
</details>
Next PR is fine.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 756 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQMInUldSXZ444fgEB4:-KQNRGBIh0xfG2xY0lta:b-w6m34c) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L756)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Note that this'll look pretty unusual to anyone new to the code (or anyone who has a grasp for on it, I guess), considering we have a `TranslateAddress` function, which, if inadequate as is, could just have an overload that accepts a BatTable and an address.
</details>
Yeah, might want to add DBAT_ONLY and IBAT_ONLY flags to `TranslateAddress` which terminate early. 

Or just create `TranslateBatAddess`, considering we already have `TranslatePageAddress`

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 778 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQMJJKy2XyikmGjXaP3:-KQNTv9QaFw-R-M0FyFf:b-q5ker5) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L778)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Same here
</details>
And could we get a comment saying why we don't optimise for page translations?

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 1232 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQNTXeNXjrXA4eVS72H:-KQNTXeNXjrXA4eVS72I:b-bej325) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L1232)):*
> ```cpp
> template <const XCheckTLBFlag flag>
> TranslateAddressResult TranslateAddress(const u32 address)
> {
> ```

Could we get a comment saying BAT translations are defined by the PowerPC spec as taking priority over page translations, just so any new coders know what is going on here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by phire** [comment]
Actually, now that I think about it...

Are `IsOptimizableMMIOAccess` and friends relying on on BAT translations of optimizable accesses to never change? Because now that we have dynamic BATs, they can change, which would require invalidating all jit blocks with optimises accesses.

**Comment by magumagu** [comment]
As you can probably guess from my lack of activity, I'm not going to be able to work on this in the near future.

The additional changes look fine.

> Are  IsOptimizableMMIOAccess  and friends relying on on BAT translations of optimizable accesses to never change?

Yes.

> Because now that we have dynamic BATs, they can change, which would require invalidating all jit blocks with optimises accesses. 

DBatUpdated() calls JitInterface::ClearSafe().  Maybe worth a comment in the code, since this point has come up more than once.

**Comment by degasus** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 756 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQMInUldSXZ444fgEB4:-KQk50_UiNFZ_SzTlc1o:b-n53gaw) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L756)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Yeah, might want to add DBAT_ONLY and IBAT_ONLY flags to `TranslateAddress` which terminate early. 
> 
> Or just create `TranslateBatAddess`, considering we already have `TranslatePageAddress`
</details>
I've created a small inline function for this lines. I'd like to cleanup TranslateAddress later on.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 778 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQMJJKy2XyikmGjXaP3:-KQk4zwbJNBeVwMhXsPO:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L778)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> And could we get a comment saying why we don't optimise for page translations?
</details>
Done.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 1232 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQNTXeNXjrXA4eVS72H:-KQk2ZYos3FfkCD7Uj7M:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L1232)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Could we get a comment saying BAT translations are defined by the PowerPC spec as taking priority over page translations, just so any new coders know what is going on here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by degasus** [comment]
Review status: 15 of 20 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 1168 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQk7f_A0uEkZXFjteDi:-KQk7f_B2jAEIh_MsOuq:b-6dh6i3) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7da4d98a034420576817c9b5fa8b6b7bb05da8bc/Source/Core/Core/PowerPC/MMU.cpp#L1168)):*
> ```cpp
>         // bit from the bottom is whether we can use the fastmem arena.
>         u32 valid_bit = 0x1;
>         if (Memory::bFakeVMEM && ((address & 0xFE000000) == 0x7E000000))
> ```

By the way, I dislike this 0x2 bit generation here a lot. I'd like to move it to IsOptimizableRAMAddress with physical_regions in mind. Do you think this shall be done now, or in a cleanup PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Comment by phire** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 756 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQMInUldSXZ444fgEB4:-KQoAE3YMXSPT27aIqUM:bfjlvh7) ([raw file](https://github.com/dolphin-emu/dolphin/blob/caeefadb218bddd4a520f88d2681c4e1c40390e4/Source/Core/Core/PowerPC/MMU.cpp#L756)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> I've created a small inline function for this lines. I'd like to cleanup TranslateAddress later on.
</details>
Yeah, that looks fine.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 1168 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQk7f_A0uEkZXFjteDi:-KQoAzZlvOFmKzvj_VkK:b-65u3wr) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7da4d98a034420576817c9b5fa8b6b7bb05da8bc/Source/Core/Core/PowerPC/MMU.cpp#L1168)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> By the way, I dislike this 0x2 bit generation here a lot. I'd like to move it to IsOptimizableRAMAddress with physical_regions in mind. Do you think this shall be done now, or in a cleanup PR?
</details>
It does seem a little weird to hardcode the physical mapping in here. 

It's your choice really. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-:-KQoC-YKTw2450aQCFcH:bnfp4nl)*

**Comment by degasus** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/PowerPC/MMU.cpp, line 1168 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146#-KQk7f_A0uEkZXFjteDi:-KQoC8dgHsbNqjd5kDJD:bz0zee) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7da4d98a034420576817c9b5fa8b6b7bb05da8bc/Source/Core/Core/PowerPC/MMU.cpp#L1168)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> It does seem a little weird to hardcode the physical mapping in here. 
> 
> It's your choice really. 
> 
</details>
We hardcode it in MMU.cpp a few times. Should be cleaned up, but I don't want to add big cleanups in the same PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4146)*

**Pull request merged**